### PR TITLE
Update logic surrounding max_vals and max_epochs

### PR DESCRIPTION
**Pull request by sleepinyourhat**
Currently, bert_adam and other optimizers that need an upper bound on total training steps only get that upper bound when max_epochs is set, and the bound can be a bit off because we won't actually stop training until a validation is hit. With this change, that bound should be set correctly using whichever is more restrictive of max_epochs or max_vals. Addresses the main issue behind #863, and should be sufficient for pretty much all use cases.

### CODE REVIEW

File: config/defaults.conf [hunk 1]
```diff
@@ -172,6 +172,13 @@ patience = 5  // Patience in early stopping. Training will stop if performance d
               // all in patience + 1 validations.
 keep_all_checkpoints = 0  // If set, keep checkpoint files from every validation. Otherwise, keep
                           // only the best and (if different) most recent.
+delete_checkpoints_when_done = 0  // If set, delete *all* saved checkpoints when the run is
```
Comment (W4ngatang) [inline config/defaults.conf]: is it really space consuming to keep just the best (or last N) checkpoints around? Are they 1GB each?

**Comment by W4ngatang** [comment]
Is the PR title accurate? Didn't see changes to `max_vals` or `max_epochs`

**Comment by sleepinyourhat** [comment]
Eek. PTAL now.

**Comment by W4ngatang** [comment]
Oh, I totally goofed and forgot to delete the task-specific optimization params when I deleted task-specific optimizers. Do you want to also blow those up in this PR?

**Comment by sleepinyourhat** [comment]
Sure. Mind pushing to this branch?

**Comment by sleepinyourhat** [comment]
Eh, mind just making an issue out of it?

**Pull request merged**